### PR TITLE
Introduce new generic ADC driver for ADS1115 chip

### PR DESCRIPTION
**Pull request by SalimTerryLi**
**Describe problem solved by this pull request**
This driver would provide ADC interface for boards without on-chip ADC function such as RPi.

**Describe your solution**
Sampling on trigger, SPS=250, VREF=6.144.
Scheduled at 80Hz, looping through its four multiplexed channel, the driver publishes a new adc_report after all 4 channels are captured, at 20Hz.

This driver is designed to run on any platforms. But ADC subscribers are not checking devID at now, so it should only run on boards without on-chip ADC.

**Test data / coverage**
Works fine on RPi.
图片

### CODE REVIEW

File: src/drivers/adc/ADS1115/ADS1115.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#ifndef PX4_ADS1115_H
```
Comment (dagar) [inline src/drivers/adc/ADS1115/ADS1115.h]: #pragma once

**Comment by dagar** [comment]
I don't quite understand the need for the I2C_Interface and the ADS1115_Wrapper. Why not just have a simple ADS1115 that extends device::I2C? 

Example: https://github.com/PX4/Firmware/blob/master/src/drivers/magnetometer/isentek/ist8308/IST8308.hpp

The other minor thing is the placement within another module (drivers/adc). I'd like to avoid modules within modules, although obviously our organization system needs work (where to put or name the generic adc driver).

**Comment by SalimTerryLi** [comment]
> I don't quite understand the need for the I2C_Interface and the ADS1115_Wrapper. Why not just have a simple ADS1115 that extends device::I2C?
> 
> Example: https://github.com/PX4/Firmware/blob/master/src/drivers/magnetometer/isentek/ist8308/IST8308.hpp
> 
> The other minor thing is the placement within another module (drivers/adc). I'd like to avoid modules within modules, although obviously our organization system needs work (where to put or name the generic adc driver).

I just want to make it more clear to understand. `ADS1115` is only a logic driver and the wrapper do the other things like device registering and scheduling. PCA9685(#14269) is also designed this way.

File: src/drivers/adc/ADS1115/ads1115_main.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+	return instance;
+}
+
+void ADS1115::print_usage()
+{
+	PRINT_MODULE_USAGE_NAME("ads1115", "driver");
+	PRINT_MODULE_USAGE_COMMAND("start");
+	PRINT_MODULE_USAGE_PARAMS_I2C_SPI_DRIVER(true, false);
+	PRINT_MODULE_USAGE_PARAM_INT('A', 0x48, 0, 0xff, "Address", true);
```
Comment (bkueng) [inline src/drivers/adc/ADS1115/ads1115_main.cpp]:
```suggestion
	PRINT_MODULE_USAGE_PARAMS_I2C_ADDRESS(0x48);
```

Comment (SalimTerryLi) [reply src/drivers/adc/ADS1115/ads1115_main.cpp]: 962ff6143

**Comment by SalimTerryLi** [comment]
updated & checked for availablity

**Comment by dagar** [comment]
Happy to merge once https://github.com/PX4/Firmware/pull/14477 is in.

**Comment by SalimTerryLi** [comment]
> 
> 
> Happy to merge once #14477 is in.

Ready.

File: boards/px4/fmu-v3/default.cmake [hunk 3]
```diff
@@ -19,6 +19,7 @@ px4_add_board(
 		TEL4:/dev/ttyS6
 	DRIVERS
 		adc/board_adc
+		adc/ADS1115
```
Comment (dagar) [inline boards/px4/fmu-v3/default.cmake:22]: This is nitpicky, but for consistency how about keeping the folder name lowercase like all the other drivers and modules?

Comment (SalimTerryLi) [reply boards/px4/fmu-v3/default.cmake]: Confirmed.

**Review by dagar — APPROVED** [review]

**Review by ibrahim0v — COMMENTED** [review]
Everything seems ok to me @SalimTerryLi great work. 
Update: ~The only issue I have faced testing your driver is when the driver is already running and is started again, it crashes and the driver can never be started again unless the autopilot is rebooted.~

**Comment by SalimTerryLi** [comment]
> Everything seems ok to me @SalimTerryLi great work. The only issue I have faced testing your driver is when the driver is already running and is started again, it crashes and the driver can never be started again unless the autopilot is rebooted.

Oh yes. It seems I forget to do ScheduleClear() on exit. I will test and fix it later. Thanks for your report.

**Comment by SalimTerryLi** [comment]
> Everything seems ok to me @SalimTerryLi great work. The only issue I have faced testing your driver is when the driver is already running and is started again, it crashes and the driver can never be started again unless the autopilot is rebooted.

Could you please provide more information? How it crashed? Anything outputted? It works just fine on my Pi. BTW there is no need to manually call `ScheduleClear();`, as it is done by I2CSPIDriverBase.
```
pxh> ads1115 start -I
WARN  [SPI_I2C] Already running on bus 1
WARN  [SPI_I2C] UnknownApp: no instance started (no device on bus?)
Command 'ads1115' failed, returned -1.
pxh> ads1115 start -I
WARN  [SPI_I2C] Already running on bus 1
WARN  [SPI_I2C] UnknownApp: no instance started (no device on bus?)
Command 'ads1115' failed, returned -1.
pxh> ads1115 INFO  [ecl/EKF] 624834243: EKF aligned, (baro hgt, IMU buf: 18, OBS buf: 14)
pxh> ads1115 stop
INFO  [ads1115] stopping
pxh> ads1115 stop
ERROR [SPI_I2C] Not running
Command 'ads1115' failed, returned -1.
pxh> ads1115 start -I
ads1115 #0 on I2C bus 1
pxh> ads1115 start -I
WARN  [SPI_I2C] Already running on bus 1
WARN  [SPI_I2C] UnknownApp: no instance started (no device on bus?)
Command 'ads1115' failed, returned -1.
pxh> ads1115 stop
INFO  [ads1115] stopping
```

**Comment by ibrahim0v** [comment]
Oh ok it is my bad. I am working on my customized firmware that is still back at 1.9.0. Therefore I have ammended your main function.

**Pull request merged**